[12] Add generated data timeseries notebook

Jack-0-0 opened the pull request:
Closes #12.

This PR:
* adds functions for generating synthetic research funding and private investment data (`iss_forecasting/analysis/utils/time_series.py`)
* adds a notebook using the generated time series data to plot charts, test for stationarity, cross correlation plots and granger causality (`iss_forecasting/analysis/tech_category_level/research_funding_and_investment_generated_data.py`)

Jack-0-0 commented:
There was an issue where `investment_raised_total` was being found to granger cause `research_funding_total`. This was due to using a random seed in the function `iss_forecasting.analysis.utils.time_series.generate_ts`.

Jack-0-0 commented:
At first, I was surprised by the results in the granger tests where for example in the yearly data, the tests would suggest that research funding granger causes investment at lag 1 (which was expected) but then also at lags 2 and 3.
This is due to how the granger tests work. When testing lag 2, it includes lag 1 and lag 2. When testing at lag 10 it is including lag 1-lag 10. The results make sense as the important lag is still included when testing the later lags.

beingkk commented:
Thank you, Jack - I will aim to take a look at this by end of Thursday.

On Wed, 9 Feb 2022 at 13:07, Jack Rasala ***@***.***> wrote:

> @Jack-0-0 <https://github.com/Jack-0-0> requested your review on: #13
> <https://github.com/nestauk/iss_forecasting/pull/13> [12] Add generated
> data timeseries notebook.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nestauk/iss_forecasting/pull/13#event-6034223774>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHDWGHKZPMA7ZL2ZPO4HJV3U2JRKNANCNFSM5NSHYA3Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

-- 


58 Victoria Embankment
London
EC4Y 0DS

				www.nesta.org.uk 
<http://www.nesta.org.uk/> |					Twitter <http://www.twitter.com/nesta_uk> 
|				 				LinkedIn 
<http://www.linkedin.com/groups?gid=1868227&goback=%2Egdr_1274367066783_1> 
|				 				Facebook 
<http://www.facebook.com/pages/NESTA/116788428355432?v=wall&ref=sgm>



If 
you no longer want to receive emails from Nesta, send an email to 
***@***.*** ***@***.***>. This email and any 
attachments are confidential and may be 				subject to legal privilege. Any 
use, copying or disclosure other 				than by the intended recipient is 
unauthorised. If you have 				received this message in error, please notify 
the sender 				immediately or by email to				 				***@***.*** 
***@***.***> and delete this message and any 				
copies from your computer and network. The views expressed in 				this 
email are those of the author and do not necessarily 				reflect the views 
of Nesta. Nesta is a company limited by 				guarantee and registered in 
England and Wales with company 				number 7706036 and charity number 
1144091. Registered as a 				charity in Scotland number SC042833. 
Registered office: 58 Victoria Embankment, London, EC4Y 0DS. 

--

beingkk commented on iss_forecasting/analysis/utils/time_series.py:
```diff
@@ -0,0 +1,76 @@
+        lag: Number of time periods to have research_funding_total lead to investment_raised_total
+        start: Start date. Defaults to "01/01/2007".
+        end: End date. Defaults to "12/31/2021".
+
```
Hey, would be great if you can describe here the `rf_scale`, `rf_const`, `factor` and `pi_scale` input variables as well.

beingkk commented on iss_forecasting/analysis/utils/time_series.py:
```diff
@@ -0,0 +1,76 @@
+    """Generate dataframe containing columns for time_period, research_funding_total,
+    investment_raised_total and tech_category. research_funding_total and
+    investment_raised_total have generated time series data, where investment_raised_total
+    is research_funding_total shifted by the lag variable.
```
Could you please add a brief explanation of how the time series are generated?
(I guess you have something like first-order autoregressive process here, but you've also injected random noise into the coefficient and constant)

beingkk commented on iss_forecasting/analysis/tech_category_level/research_funding_and_investment_generated_data.py:
```diff
@@ -0,0 +1,195 @@
+
+# %%
+# Generate monthly data
+gen_ts_months = generate_ts(period="M", lag=12)
```
You could add here a random seed variable, in order to make the results reproducible.

beingkk commented:
Thank you @Jack-0-0 - I've left a few small comments, but otherwise looks good to merge! 

Happy to see that the measures behave pretty much as expected. Perhaps would be interesting to also inspect time series with some underlying trend (not just constant/noisy baseline) but can leave that for some other time to experiment with.

Jack-0-0 commented:
> Thank you @Jack-0-0 - I've left a few small comments, but otherwise looks good to merge!
> 
> Happy to see that the measures behave pretty much as expected. Perhaps would be interesting to also inspect time series with some underlying trend (not just constant/noisy baseline) but can leave that for some other time to experiment with.

Thank you for the review @beingkk. I've updated `generate_ts`'s docstring and added a random seed.
